### PR TITLE
Ctw 476/ccd condition history viewer

### PR DESCRIPTION
**Pull request by pmadria3**
This PR helps to incorporate the CCD into the condition history page to help provide source of truth and additional information for the conditions if there is a CCD document for the condition. If there is no CCD document associated with the file then no document will show up.

https://user-images.githubusercontent.com/98342697/209174468-585c39a1-374e-43d9-8283-bfa4dcc2a86f.mov

### CODE REVIEW

File: src/components/content/CCDA/ccda_viewer/index.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import { DocumentDetails } from "./components/DocumentDetails";
+import { Header } from "./components/Header/Header";
+import { Section } from "./components/Section/Section";
+import { DocumentOnlyProps, SectionType } from "./types";
+
+const useStyles = makeStyles(() => ({
+  container: {
+    width: "calc(100vw - 200px - 24px - 24px - 16px - 16px - 2px)",
+  },
+}));
```
Comment (mrosata) [inline src/components/content/CCDA/ccda_viewer/index.tsx:12]: We probably shouldn't add `@material-ui/core` _(ie: `makeStyles()`)_ to the project.

These can just be refactored out into SCSS and/or with the `classnames` library

File: src/components/content/CCDA/ccda_viewer/helpers/parseToISOString.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            result += `${timezoneSign}${timezoneHour}`;
+            if (isCorrectValue(timezoneMinute)) result += `:${timezoneMinute}`;
+            else result += ":00";
+          } else result += "-00:00";
+        } else result += "T00:00:00-00:00";
+      }
+    }
+  }
+  return result;
+};
```
Comment (mrosata) [inline src/components/content/CCDA/ccda_viewer/helpers/parseToISOString.ts:100]: Even though this was probably copied from fhirplace and we may know that it works, it would be better to use a date library to parse and build date stings. The one that is already installed is [`date-fns`](https://date-fns.org/).

> `date-fns` depends on `moment.js` which is a deprecated library as of a few years ago, eventually we may want to migrate off `date-fns`, but that's for another day.

File: src/components/content/CCDA/ccda_viewer/helpers/parseMany.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { isEmpty } from "lodash";
+
+export function parseMany<T>(
+  parser: (el: Document) => T,
+  data: Document[]
+): T[] {
+  if (isEmpty(data)) return [];
+
+  return data.map((val) => parser(val));
+}
```
Comment (mrosata) [inline src/components/content/CCDA/ccda_viewer/helpers/parseMany.ts:10]: This file isn't needed. Assuming this was added because:
```ts
const data = null;

data.map(parser);  // Throws
```
would throw an error because `.map` only works on Arrays.. but lodash `_.map` already accounts for this. So all the instances of parseMany can be replaced with lodash map.

An example:
```ts
      given: parseMany<string>(
        (n) => String(xpath.select1("string(text())", n)),
        xpath.select("(*[name()='given'])", nameXml) as Document[]
      ),
```
Would refactor into:
```ts
import { map } from "lodash/fp";

      given: map(
        (n) => `${xpath.select1("string(text())", n)}`,
        xpath.select("(*[name()='given'])", nameXml)
      ),
```
or if using "lodash" instead of "lodash/fp"
```ts
import { map } from "lodash;

      given: map(
        xpath.select("(*[name()='given'])", nameXml),
        (n) => `${xpath.select1("string(text())", n)}`
      ),
```

File: package.json [hunk 4]
```diff
@@ -103,14 +104,22 @@
   "dependencies": {
     "@headlessui/react": "^1.7.3",
     "@heroicons/react": "^1.0.6",
+    "@material-ui/core": "^4.12.4",
```
Comment (luthfib) [inline package.json:107]: I think we if we can avoid installing material-ui/core, which seeing the surface area it is not that big, we should. This is quite a big library, and we dont use it much. Either in this pr or a additional ticket.

File: src/fhir/conditions.ts [hunk 5]
```diff
@@ -158,3 +162,84 @@ function filterAndSort(conditions: ConditionModel[]): ConditionModel[] {
...
+) => {
+  const practitioner = await getPractitioner(practitionerId, requestContext);
+  const display = practitioner.fullName;
+
+  return {
+    reference: `Practitioner/${practitionerId}`,
+    type: "Practitioner",
+    display,
+  };
+};
```
Comment (luthfib) [inline src/fhir/conditions.ts:178]: Why did you add this? I don't see it being used anywhere

File: src/components/core/modal-ccda.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  useEffect(() => {
+    if (rawBinary.isBinary) {
+      setIsXMLData(true);
+    }
+  }, [isXMLData, rawBinary.isBinary]);
+
+  return (
+    <Modal {...modalProps}>
+      {isXMLData && rawBinary.xmlBinary?.data && rawBinary.xmlBinary.data && (
```
Comment (luthfib) [inline src/components/core/modal-ccda.tsx:27]: why do we have `rawBinary.xmlBinary?.data && rawBinary.xmlBinary.data`. Seems you are checking the same thing twice

File: src/components/content/conditions-history.tsx [hunk 7]
```diff
@@ -180,17 +238,49 @@ export function ConditionHistory({
...
+                    <DocumentButton
+                      onClick={() => {
+                        setIsModalOpen(true);
+                        setRawBinary(idMap.get(entry.id));
+                      }}
+                    />
+                  )}
+                </>
+              ),
+            }))}
```
Comment (luthfib) [inline src/components/content/conditions-history.tsx:262]: This is really confusing use of children. It just display next to the button. Please change this to be something a bit more indicative of what it is actually doing. Might need to change it in Details as well. This prop makes me think I would be changing the actual content in the component.

File: src/components/content/conditions-history.tsx [hunk 8]
```diff
@@ -141,6 +162,29 @@ export function ConditionHistory({
         setConditionsWithDate(conditionsDataDeduped.filter((d) => d.date));
         setConditionsWithoutDate(conditionsDataDeduped.filter((d) => !d.date));
         setLoading(false);
+
+        // Binary Doc stuff
+        const requestContext = await getRequestContext();
+
+        const turnConditiontoDataList = setupData(condition);
```
Comment (luthfib) [inline src/components/content/conditions-history.tsx:169]: `turnConditiontoDataList` is a confusing name. Still not quite sure what it is

File: src/components/content/conditions-history.tsx [hunk 9]
```diff
@@ -141,6 +162,29 @@ export function ConditionHistory({
...
+        );
+
+        setIDMap(binaryDocs);
+
+        if (binaryDocs.get(condition.id)?.isBinary) {
+          setIsBinaryDocumentForOriginalEntry(true);
+          setRawBinary(binaryDocs.get(condition.id));
+        }
+
+        setLoadingSourceDocument(false);
```
Comment (luthfib) [inline src/components/content/conditions-history.tsx:187]: This whole block could probably be a function in itself.

File: src/components/content/conditions-history.tsx [hunk 10]
```diff
@@ -107,10 +115,23 @@ export function ConditionHistory({
...
     useState<CollapsibleDataListStackEntries>([]);
   const [loading, setLoading] = useState(true);
+  const [loadingSourceDocument, setLoadingSourceDocument] = useState(true);
   const [conditionForSearch, setConditionForSearch] =
     useState<ConditionModel>();
   const historyResponse = useConditionHistory(conditionForSearch);
+  const { getRequestContext } = useCTW();
+
+  const [
+    isBinaryDocumentForOriginalEntry,
```
Comment (luthfib) [inline src/components/content/conditions-history.tsx:125]: What is `isBinaryDocumentForOriginalEntry`. What is original entry?

File: src/components/content/CCDA/ccda_viewer/components/Custodian/data.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+export const getCustodianData = (
+  document: Document
+): GeneralInfo | undefined => {
+  const representedCustodianOrganization = xpath.select1(
+    "*[name()='ClinicalDocument']/*[name()='custodian']/*[name()='assignedCustodian']/*[name()='representedCustodianOrganization']",
+    document
+  ) as Document;
+
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!representedCustodianOrganization) return undefined;
```
Comment (luthfib) [inline src/components/content/CCDA/ccda_viewer/components/Custodian/data.ts:14]: I think adding   `// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition` is actually wrong. The reason why it thinks it is not necessary is because you cast `representedCustodianOrganization` as Document. If xpath can return undefined then you could also just cast `as Document | undefined`

**Review by luthfib — COMMENTED** [review]
Seems we have a lot of `// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition` in the code that was ported over. In almost all instances it is incorrect use of it. Not sure how worth it is to fix some of those since we will be maintaining this code.

**Comment by pmadria3** [comment]
Moving storybook to follow up 2.0 ticket

File: src/components/content/CCDA/ccda_viewer/components/Authorization/Authorization.tsx [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import { isEmpty } from "lodash";
+import { DocumentOnlyProps } from "../../types";
+import { Table } from "../Table/Table";
+import { getAuthorizationData } from "./data";
+
+export const Authorization = ({ document }: DocumentOnlyProps) => {
+  const authorizations = getAuthorizationData(document);
+
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!authorizations || isEmpty(authorizations)) return null;
```
Comment (luthfib) [inline src/components/content/CCDA/ccda_viewer/components/Authorization/Authorization.tsx:10]: Why are we checking the same thing twice? We are checking if it is not empty and then checking isEmpty right after it. Seems redundant?

File: src/components/content/CCDA/ccda_viewer/components/InFulfillmentOf/InFulfillmentOf.tsx [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import { isEmpty } from "lodash";
+import { DocumentOnlyProps } from "../../types";
+import { Table } from "../Table/Table";
+import { getInFulfillmentOfData } from "./data";
+
+export const InFulfillmentOf = ({ document }: DocumentOnlyProps) => {
+  const inFulfillmentOfs = getInFulfillmentOfData(document);
+
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!inFulfillmentOfs || isEmpty(inFulfillmentOfs)) return null;
```
Comment (luthfib) [inline src/components/content/CCDA/ccda_viewer/components/InFulfillmentOf/InFulfillmentOf.tsx]: Same here as the comment above

File: src/components/content/CCDA/ccda_viewer/helpers/getHumanName.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  R: "Religious",
+  SNDX: "Soundex",
+  SRCH: "Search",
+  SYL: "Syllabic",
+};
+
+export const getHumanName = (name: Document | Document[]): string => {
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!name) return "";
+
```
Comment (luthfib) [inline src/components/content/CCDA/ccda_viewer/helpers/getHumanName.ts:25]: Remove   `// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition` and make name a optional param

File: src/components/content/CCDA/ccda_viewer/helpers/getPeriod.ts [hunk 15]
```diff
@@ -0,0 +1,46 @@
+import { isBefore, isEqual } from "date-fns";
+import { Period } from "fhir/r4";
+import xpath from "xpath";
+import { ccdaDatetimeToISO, displayDateTimeasString } from "@/fhir/formatters";
+
+export const getPeriod = (xmlData: Document): Period => {
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!xmlData) return {};
```
Comment (luthfib) [inline src/components/content/CCDA/ccda_viewer/helpers/getPeriod.ts:8]: Make `xmlData` optional if we are going to have that check

File: src/App.tsx [hunk 16]
```diff
@@ -16,7 +16,7 @@ const {
   VITE_AUTH0_CLIENT_ID,
   VITE_AUTH0_AUDIENCE,
   VITE_AUTH0_CALLBACK_PATH,
-  VITE_ENV = "dev",
+  VITE_ENV = "sandbox",
```
Comment (luthfib) [inline src/App.tsx:19]: Don't change this

Comment (pmadria3) [reply src/App.tsx:19]: good catch

File: src/fhir/conditions.ts [hunk 17]
```diff
@@ -174,3 +177,70 @@ function filterAndSort(conditions: ConditionModel[]): ConditionModel[] {
...
+              headers: {
+                Authorization: `Bearer ${requestContext.authToken}`,
+                accept: "application/json",
+                "Content-Type": "application/json+fhir",
+              },
+            });
+            documentBinary = await response.json();
+          }
+        }
+      }
```
Comment (luthfib) [inline src/fhir/conditions.ts:235]: I am pretty sure we can change this to....

```suggestion
 if (conditionsJSON.entry[0].resource.entity[0].role === "source") {
        const binaryID =
          conditionsJSON.entry[0].resource.entity[0].what.reference;

        const endpointBinaryUrl = `${getZusApiBaseUrl(
          requestContext.env
        )}/fhir/${binaryID}`;

        const response = await fetch(endpointBinaryUrl, {
          headers: {
            Authorization: `Bearer ${requestContext.authToken}`,
            accept: "application/json",
            "Content-Type": "application/json+fhir",
          },
        });
        documentBinary = await response.json();
      }
```

File: src/components/content/conditions-history.tsx [hunk 18]
```diff
@@ -183,14 +237,42 @@ export function ConditionHistory({
...
+                        <DocumentButton
+                          onClick={() => {
+                            setIsModalOpen(true);
+                            setRawBinary(idMap.get(entry.id));
+                          }}
+                        />
+                      )}
+                    </>
+                  ),
+                }))}
```
Comment (luthfib) [inline src/components/content/conditions-history.tsx:275]: Can we make a component for this? They seem almost identical

**Review by luthfib — APPROVED** [review]

**Review by mrosata — DISMISSED** [review]
